File: src/api/endpoints/auth/AuthEndpoint.ts
import { Logger } from "ts-log";

import { HeaderProvider } from "../../../Func";
import ApiToken from "../../../models/ApiToken";
import ApiResponse from "../../ApiResponse";
import ApiEndpoint from "../ApiEndpoint";
import TokenManager from "../auth/TokenManager";
import DefaultApiConfig from "../DefaultApiConfig";
import CreateApiTokenRequest from "./request/CreateApiTokenRequest";
import SignOutRequest from "./request/SignOutRequest";

/**
 * Endpoints for interacting with auth.
 * */
export default class AuthEndpoint extends ApiEndpoint {
  constructor(config: DefaultApiConfig, tokenManager: TokenManager, log?: Logger, headerProvider?: HeaderProvider) {
    super(config, tokenManager, log, headerProvider);
  }

  protected override endpoint(): string {
    return "auth";
  }

  /**
   * Invalidate a refresh token.
   *
   * @param oldToken - Token to be invalidated
   *
   * @virtual
   * */
  public async signOut(oldToken: SignOutRequest): Promise<ApiResponse> {
    return await this.postRequest<string>("/signOut", oldToken);
  }

  /**
   * Gets all api tokens for the user.
   *
   * */
  public async getApiTokens(): Promise<ApiResponse<ApiToken[]>> {
    return await this.getRequest<ApiToken[]>(`/apiTokens`);
  }

  /**
   * Creates an api token.
   *
   * @param newToken - Emergency details for the new emergency
   * @returns The newly-created api token
   *
   * @virtual
   * */
  public async createApiToken(newToken: CreateApiTokenRequest): Promise<ApiResponse<string>> {
    return await this.postRequest<string>("/apiTokens", {
      name: newToken.name,
      expirationDate: newToken.expirationDate?.toISOString(),
    });
  }

  /**
   * Delete an api token.
   *
   * @param id - The id of the api token to delete
   *
   * */
  public async deleteApiToken(id: string): Promise<ApiResponse> {
    return await this.deleteRequest(`/apiTokens/${id}`);
  }
}

File: src-dotnet/api/endpoints/auth/AuthEndpoint.cs
namespace MedrunnerApiClient.Api.Endpoints.Auth
{
    /// <summary>
    /// Represents the authentication endpoint.
    /// </summary>
    public class AuthEndpoint : ApiEndpoint
    {
        public AuthEndpoint(ApiConfig config, TokenManager tokenManager, ILogger logger)
            : base(config, tokenManager, logger)
        {
        }

        protected override string Endpoint => "auth";

        /// <summary>
        /// Invalidate a refresh token.
        /// </summary>
        /// <param name="oldToken">Token to be invalidated</param>
        public async Task<ApiResponse<string>> SignOutAsync(SignOutRequest? oldToken = null)
        {
            return await PostRequestAsync<string>("/signOut", oldToken);
        }

        /// <summary>
        /// Gets all API tokens for the user.
        /// </summary>
        public async Task<ApiResponse<List<ApiToken>>> GetApiTokensAsync()
        {
            return await GetRequestAsync<List<ApiToken>>("/apiTokens");
        }

        /// <summary>
        /// Creates an API token.
        /// </summary>
        /// <param name="newToken">Details for the new API token</param>
        /// <returns>The newly-created API token</returns>
        public async Task<ApiResponse<string>> CreateApiTokenAsync(CreateApiTokenRequest newToken)
        {
            var payload = new
            {
                name = newToken.Name,
                expirationDate = newToken.ExpirationDate?.ToString("o")
            };
            return await PostRequestAsync<string>("/apiTokens", payload);
        }

        /// <summary>
        /// Delete an API token.
        /// </summary>
        /// <param name="id">The id of the API token to delete</param>
        public async Task<ApiResponse<string>> DeleteApiTokenAsync(string id)
        {
            return await DeleteRequestAsync<string>($"/apiTokens/{id}");
        }
    }
}
